File: 1248/solution.py
class Solution:
    def numberOfSubarrays(self, nums: list[int], k: int) -> int:
        l = 0
        mid = 0
        odds = 0
        res = 0
        for r in range(len(nums)):
            odds += nums[r] % 2
            if odds > k:
                mid += 1
                l = mid
                odds -= 1

            while odds == k and nums[mid] % 2 == 0:
                mid += 1

            if odds == k:
                res += mid - l + 1
        return res

File: 1248/solution.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        int l = 0;
        int mid = 0;
        int odds = 0;
        int res = 0;
        for (int r = 0; r < nums.length; r++) {
            odds += nums[r] % 2;
            if (odds > k) {
                mid++;
                l = mid;
                odds--;
            }
            while (odds == k && nums[mid] % 2 == 0) {
                mid++;
            }
            if (odds == k) {
                res += mid - l + 1;
            }
        }
        return res;
    }
}

// time O(n)
// space O(1)